pkg/benchmark/throughput.go

```diff
@@ -0,0 +1,191 @@
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package benchmark
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/gomega"
+	"github.com/submariner-io/shipyard/test/e2e/framework"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/tests"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var Verbose bool
+
+func TestThroughput(args []string, restConfigProducer restconfig.Producer, intraCluster bool) {
+	err := tests.ConfigureFramework(args, restConfigProducer)
+	if err != nil {
+		fmt.Println(err.Error())
+
+		return
+	}
+
+	if Verbose {
+		fmt.Printf("Performing throughput tests\n")
+	}
+
+	startThroughputTests(intraCluster)
+}
+
+func startThroughputTests(intraCluster bool) {
+	var f *framework.Framework
+
+	gomega.RegisterFailHandler(func(message string, callerSkip ...int) {
+		if f != nil {
+			cleanupFramework(f)
+		} else {
+			framework.RunCleanupActions()
+		}
+		panic(message)
+	})
+
+	f = initFramework("throughput")
+
+	clusterAName := framework.TestContext.ClusterIDs[framework.ClusterA]
+
+	if !intraCluster {
+		testParams := benchmarkTestParams{
+			ClientCluster:       framework.ClusterA,
+			ServerCluster:       framework.ClusterB,
+			ServerPodScheduling: framework.GatewayNode,
+			ClientPodScheduling: framework.GatewayNode,
+		}
+
+		clusterBName := framework.TestContext.ClusterIDs[framework.ClusterB]
+		fmt.Printf("Performing throughput tests from Gateway pod on cluster %q to Gateway pod on cluster %q\n",
+			clusterAName, clusterBName)
+		runThroughputTest(f, testParams)
+
+		testParams.ServerPodScheduling = framework.NonGatewayNode
+		testParams.ClientPodScheduling = framework.NonGatewayNode
+
+		fmt.Printf("Performing throughput tests from Non-Gateway pod on cluster %q to Non-Gateway pod on cluster %q\n",
+			clusterAName, clusterBName)
+		runThroughputTest(f, testParams)
+	} else {
+		testIntraClusterParams := benchmarkTestParams{
+			ClientCluster:       framework.ClusterA,
+			ServerCluster:       framework.ClusterA,
+			ServerPodScheduling: framework.GatewayNode,
+			ClientPodScheduling: framework.NonGatewayNode,
+		}
+
+		fmt.Printf("Performing throughput tests from Non-Gateway pod to Gateway pod on cluster %q\n", clusterAName)
+		runThroughputTest(f, testIntraClusterParams)
+	}
+
+	cleanupFramework(f)
+}
+
+var By = func(str string) {
+	if Verbose {
+		fmt.Println(str)
```

Comment:
I think we should let the caller pass in an `io.Writer` and use `fmt.Fprintln` here instead of assuming `os.Stdout`.

Reply:
All `subctl` output is `os.Stdout`. We can add a flag to the whole `benchmark` cmd if we want to store the output someplace else.

Reply:
I was referring to making the output destination configurable in the library by passing an `io.Writer`. But that would only redirect calls to `By` - we'd also have to configure the gomega fail handler. We can do that later if need be.

Reply:
@Jaanki, since this is not under `cmd` we can’t consider it to be `subctl`-specific, so the same rules apply as elsewhere: no output unless to something controller-provided (a reporter or a writer), in fact no I/O in general unless to something controller-provided.